modules/ducktests/tests/ignitetest/tests/two_phased_rebalanced_test.py

```diff
@@ -0,0 +1,278 @@
+from ignitetest.services.utils.util import copy_file_to_dest
+from ignitetest.utils import ignite_versions
+from ignitetest.utils.ignite_test import IgniteTest
+from ignitetest.utils.version import IgniteVersion, DEV_BRANCH, LATEST_2_9, LATEST_2_8
+
+NUM_NODES_CELL = 4
+
+NUM_CELL = 2
+
+ATTRIBUTE = "CELL"
+
+CACHE_NAME = "test-cache"
+
+
+# pylint: disable=W0223
+class TwoPhasedRebalancedTest(IgniteTest):
+    """
+    Two-phase rebalancing test case.
+    """
+    # pylint: disable=R0914
+    @cluster(num_nodes=(NUM_NODES_CELL * NUM_CELL) + 1)
+    @ignite_versions(str(DEV_BRANCH), str(LATEST_2_9), str(LATEST_2_8))
+    def two_phased_rebalancing_test(self, ignite_version):
+        """
+        Test case of two-phase rebalancing.
+        Preparations.
+            1. Start 2 cells.
+            2. Load data to cache with the mentioned above affinity function.
+            3. Delete 80% of data and measure PDS size on all nodes.
+        Phase 1.
+            1. Stop two nodes in each cell, total a half of all nodes and clean PDS.
+            2. Start cleaned node with preservance of consistent id and cell attributes.
+            3. Wait for the rebalance to complete.
+        Phase 2.
+            Run steps 1-3 of Phase 1 on the other half of the cluster.
+        Verifications.
+            1. Check that PDS size reduced (compare to step 3)
+            2. Check data consistency (idle_verify --dump)
+        """
+        config = IgniteConfiguration(version=IgniteVersion(ignite_version),
+                                     data_storage=DataStorageConfiguration(
+                                         default=DataRegionConfiguration(persistent=True), checkpoint_frequency=30000),
+                                     caches=[CacheConfiguration(
+                                         name=CACHE_NAME, backups=2, affinity=Affinity(),
+                                         indexed_types=['java.lang.Long', 'byte[]'])],
+                                     metric_exporter='org.apache.ignite.spi.metric.jmx.JmxMetricExporterSpi')
+
+        # Start 2 cells.
+        cells = self.start_cells(config)
+
+        control_utility = ControlUtility(cells[0])
+        control_utility.activate()
+
+        client_config = IgniteConfiguration(client_mode=True,
+                                            version=IgniteVersion(ignite_version),
+                                            discovery_spi=from_ignite_cluster(cells[0]))
+
+        # Load data to cache.
+        app = IgniteApplicationService(
+            self.test_context,
+            client_config,
+            java_class_name="org.apache.ignite.internal.ducktest.tests.snapshot_test.DataLoaderApplication",
+            params={"start": 0,
+                    "cacheName": "test-cache",
+                    "interval": 500 * 1024,
+                    "valueSizeKb": 1}
+        )
+        app.run()
+        app.free()
```

Comment:
Could you please provide a ref to implementation of the free() method? I can't find one.

Reply:
ducktape/cluster/cluster.py
    def free(self, nodes):
        """Free the given node or list of nodes"""
        if isinstance(nodes, collections.Iterable):
            for s in nodes:
                self.free_single(s)
        else:
            self.free_single(nodes)

Reply:
I saw this code, I was looking for this code:

https://github.com/confluentinc/ducktape/blob/master/ducktape/cluster/json.py#L122

I'm OK then.